core/finality-grandpa/src/communication.rs

```diff
@@ -0,0 +1,290 @@
+		// when locals exist, sign messages on import
+		if let Some((ref pair, local_id)) = self.locals {
+			let encoded = localized_payload(self.round, self.set_id, &msg);
+			let signature = pair.sign(&encoded[..]);
+			let signed = SignedMessage::<Block> {
+				message: msg,
+				signature,
+				id: local_id,
+			};
+
+			// forward to network and to inner sender.
+			self.network.send_message(self.round, self.set_id, signed.encode());
+			let _ = self.sender.unbounded_send(signed);
+		}
+
+		Ok(AsyncSink::Ready)
+	}
+
+	fn poll_complete(&mut self) -> Poll<(), Error> { Ok(Async::Ready(())) }
+
+	fn close(&mut self) -> Poll<(), Error> {
+		// ignore errors since we allow this inner sender to be closed already.
+		match self.sender.close() {
+			Ok(Async::NotReady) => Ok(Async::NotReady),
+			_ => Ok(Async::Ready(()))
+		}
+	}
+}
+
+impl<Block: BlockT, N: Network> Drop for OutgoingMessages<Block, N> {
+	fn drop(&mut self) {
+		self.network.drop_messages(self.round, self.set_id);
+	}
+}
+
+/// A sink for outgoing messages. This signs the messages with the key,
+/// if we are an authority. A stream for the signed messages is also returned.
+///
+/// A future can push unsigned messages into the sink. They will be automatically
+/// broadcast to the network. The returned stream should be combined with other input.
+pub(crate) fn outgoing_messages<Block: BlockT, N: Network>(
+	round: u64,
+	set_id: u64,
+	local_key: Option<Arc<ed25519::Pair>>,
+	voters: Arc<HashMap<AuthorityId, u64>>,
+	network: N,
+) -> (
+	impl Stream<Item=SignedMessage<Block>,Error=Error>,
+	impl Sink<SinkItem=Message<Block>,SinkError=Error>,
+) {
+	let locals = local_key.and_then(|pair| {
+		let public = pair.public();
+		let id = AuthorityId(public.0);
+		if voters.contains_key(&id) {
+			Some((pair, id))
+		} else {
+			None
+		}
+	});
+
+	let (tx, rx) = mpsc::unbounded();
+	let outgoing = OutgoingMessages::<Block, N> {
+		round,
+		set_id,
+		network,
+		locals,
+		sender: tx,
+	};
+
+	let rx = rx.map_err(move |()| Error::Network(
+		format!("Failed to receive on unbounded receiver for round {}", round)
+	));
+
+	(rx, outgoing)
+}
+
+fn check_compact_commit<Block: BlockT>(
+	msg: CompactCommit<Block>,
+	voters: &HashMap<AuthorityId, u64>,
+	round: u64,
+	set_id: u64,
+) -> Option<CompactCommit<Block>> {
+	use grandpa::Message as GrandpaMessage;
+	if msg.precommits.len() != msg.auth_data.len() || msg.precommits.is_empty() {
+		debug!(target: "afg", "Skipping malformed compact commit");
+		return None;
+	}
+
+	// check signatures on all contained precommits.
+	for (precommit, &(ref sig, ref id)) in msg.precommits.iter().zip(&msg.auth_data) {
+		if !voters.contains_key(id) {
+			debug!(target: "afg", "Skipping commit containing unknown voter {}", id);
+			return None;
+		}
+
+		let res = check_message_sig::<Block>(
+			&GrandpaMessage::Precommit(precommit.clone()),
+			id,
+			sig,
+			round,
+			set_id,
+		);
+
+		if let Err(()) = res {
```

Comment:
```suggestion
		if res.is_err() {
```

Reply:
I prefer to be a bit more explicit with the `()`